pkg/fqdn/dnsproxy/proxy.go

```diff
@@ -1123,23 +1122,26 @@ func shouldCompressResponse(request, response *dns.Msg) bool {
 // GeneratePattern takes a set of l7Rules and returns a regular expression pattern for matching the
 // provided l7 rules.
 func GeneratePattern(l7Rules *policy.PerSelectorPolicy) (pattern string) {
-	if l7Rules == nil {
-		l7Rules = &policy.PerSelectorPolicy{L7Rules: api.L7Rules{DNS: []api.PortRuleDNS{{MatchPattern: "*"}}}}
+	if l7Rules == nil || len(l7Rules.DNS) == 0 {
+		return matchpattern.MatchAllAnchoredPattern
 	}
 	reStrings := make([]string, 0, len(l7Rules.DNS))
 	for _, dnsRule := range l7Rules.DNS {
 		if len(dnsRule.MatchName) > 0 {
 			dnsRuleName := strings.ToLower(dns.Fqdn(dnsRule.MatchName))
-			reStrings = append(reStrings, "("+matchpattern.ToRegexp(dnsRuleName)+")")
+			dnsRuleNameAsRE := matchpattern.ToUnAnchoredRegexp(dnsRuleName)
+			reStrings = append(reStrings, dnsRuleNameAsRE)
 		}
 		if len(dnsRule.MatchPattern) > 0 {
 			dnsPattern := matchpattern.Sanitize(dnsRule.MatchPattern)
-			dnsPatternAsRE := matchpattern.ToRegexp(dnsPattern)
-			reStrings = append(reStrings, "("+dnsPatternAsRE+")")
+			dnsPatternAsRE := matchpattern.ToUnAnchoredRegexp(dnsPattern)
+			if dnsPatternAsRE == matchpattern.MatchAllUnAnchoredPattern {
+				return matchpattern.MatchAllAnchoredPattern
+			}
```

Comment:
In an offline sync, you mentioned this was awkward. Could you explain why in a code comment?

Reply:
👋 

Yeah, this line effectively makes the `MatchAllUnAnchoredPattern` var only used here. Was thinking about making the `ToUnAnchoredRegexp` func return `string, bool` and `true` when its a wildcard, but this is just as verbose and easy reason about. I don't feel too much for one over the other.

Any thoughts?

Reply:
Ah, seems straightforward enough as-is.